src/lib.rs

```diff
@@ -8,38 +8,113 @@
 //! This is done via the [`numpy`] crate which enables passing data from Rust to Python.
+//!
+//! # Examples
+//!
+//! Read a GeoTIFF file retrieved via the [`object_store`] crate, and pass it into the
+//! [`read_geotiff`](./io/geotiff/fn.read_geotiff.html) function.
+//!
+//! ```rust
+//! use std::io::Cursor;
+//!
+//! use bytes::Bytes;
+//! use cog3pio::io::geotiff::read_geotiff;
+//! use ndarray::Array2;
+//! use object_store::path::Path;
+//! use object_store::{parse_url, GetResult, ObjectStore};
+//! use tokio;
+//! use url::Url;
+//!
+//! #[tokio::main]
+//! async fn main() {
+//!     let cog_url: &str =
+//!         "https://github.com/cogeotiff/rio-tiler/raw/6.4.0/tests/fixtures/cog_nodata_nan.tif";
+//!     let tif_url: Url = Url::parse(cog_url).unwrap();
+//!     let (store, location): (Box<dyn ObjectStore>, Path) = parse_url(&tif_url).unwrap();
+//!
+//!     let stream: Cursor<Bytes> = {
+//!         let result: GetResult = store.get(&location).await.unwrap();
+//!         let bytes: Bytes = result.bytes().await.unwrap();
+//!         Cursor::new(bytes)
+//!     };
+//!
+//!     let arr: Array2<f32> = read_geotiff(stream).unwrap();
+//!     assert_eq!(arr.dim(), (549, 549));
+//!     assert_eq!(arr[[500, 500]], 0.13482364);
+//! }
+//! ```
 
 /// Modules for handling Input/Output of GeoTIFF data
 pub mod io;
 
-use std::fs::File;
+use std::io::Cursor;
 
+use bytes::Bytes;
 use ndarray::Dim;
 use numpy::{PyArray, ToPyArray};
+use object_store::{parse_url, ObjectStore};
+use pyo3::exceptions::{PyBufferError, PyFileNotFoundError, PyValueError};
 use pyo3::prelude::{pyfunction, pymodule, PyModule, PyResult, Python};
-use pyo3::wrap_pyfunction;
+use pyo3::{wrap_pyfunction, PyErr};
+use url::Url;
 
 /// Read a GeoTIFF file from a path on disk into an ndarray
 ///
 /// Parameters
 /// ----------
 /// path : str
-///     The path to the file.
+///     The path to the file, or a url to a remote file.
 ///
 /// Returns
 /// -------
 /// array : np.ndarray
 ///     2D array containing the GeoTIFF pixel data.
+///
+/// Examples
+/// --------
+/// from cog3pio import read_geotiff
+///
+/// array = read_geotiff("https://github.com/pka/georaster/raw/v0.1.0/data/tiff/float32.tif")
+/// assert array.shape == (20, 20)
 #[pyfunction]
 #[pyo3(name = "read_geotiff")]
 fn read_geotiff_py<'py>(
     path: &str,
     py: Python<'py>,
 ) -> PyResult<&'py PyArray<f32, Dim<[usize; 2]>>> {
-    // Open TIFF file from path
-    let file = File::open(path).expect("Cannot find GeoTIFF file");
+    // Parse URL into ObjectStore and path
+    let file_or_url = match Url::from_file_path(path) {
+        // Parse local filepath
+        Ok(filepath) => filepath,
+        // Parse remote URL
+        Err(_) => Url::parse(path)
+            .map_err(|_| PyValueError::new_err(format!("Cannot parse path: {path}")))?,
+    };
+    let (store, location) = parse_url(&file_or_url)
+        .map_err(|_| PyValueError::new_err(format!("Cannot parse url: {file_or_url}")))?;
```

Comment:
> ## [CodSpeed Performance Report](https://codspeed.io/weiji14/cog3pio/branches/object_store)
> ### Merging #5 will **degrade performances by 32.21%**
> 
> Comparing `object_store` ([33f4a73](https://github.com/weiji14/cog3pio/commit/33f4a73e56845a46fe616e9076d688f0360c4b3c)) with `main` ([0c37e32](https://github.com/weiji14/cog3pio/commit/0c37e32981d4a403eff1e74ec5010872668128bd))
> ### Summary
> 
> `❌ 1 (👁 1)` regressions
> 
> `🆕 1` new benchmarks
> ### Benchmarks breakdown
> 	Benchmark 	`main` 	`object_store` 	Change
> 👁 	`test_read_geotiff_local` 	423 µs 	624 µs 	-32.21%
> 🆕 	`test_read_geotiff_remote` 	N/A 	49 ms 	N/A

Acknowledging the slowdown in reading local GeoTIFF files as acceptable for now since it is only ~200µs slower, but could revisit this next time if we should have separate branches for remote files (read using `object_store`) and local files (read using `std::fs::File`).